electrum: add patch for aiorpcx 0.24 compatibility

Yarny0 opened the pull request:
`electrum` build [broke](https://hydra.nixos.org/build/287723280) due to the `aiorpcx` update in https://github.com/NixOS/nixpkgs/commit/4bbf5300d44ee9420a59650ba668f70e3f101bde via https://github.com/NixOS/nixpkgs/pull/375144 .

Upstream already has [a patch](https://github.com/spesmilo/electrum/commit/171aa5ee5ad4e25b9da10f757d9d398e905b4945), however, it&#39;s not so easy to apply:

*   the patch patches the file `/run_electrum` in the source repo
*   in the source repo, there is a symlink `/electrum/electrum` pointing to `../run_electrum`, i.e., to the file that is patched
*   in the source distribution tarball used by nixpkgs, `/electrum/electrum` is a simple copy of `/run_electrum` (probably the result of whatever process generates the tarball)
*   the file `/electrum/electrum` is used as executable in the package output
*   besides, the patch also attempts to patch the file `requirements.txt`, but fails with conflicts

Hence, although the patch will likely fix the next electrum release, just applying it to our source does not fix the issue.  To work around that, this pull request ~~also adds a copy command to the `postFixup` section to copy the patched `/run_electrum` over the unpatched `/electrum/electrum`.  This is also explained in a comment near the `fetchpatch` line, so whoever removes the patch when it is no longer needed is also reminded of removing the copy command below.~~ applies the changes from the patch file "by hand", with `substituteInPlace`.  Additionally, it converts existing `substituteInPlace` invocations to `--replace-fail` and drop two unneeded replacements (see the commits for details).

Note: https://github.com/NixOS/nixpkgs/pull/309576 also adds a (also temporary) patch for `aiorpcx` compatibility and experienced the same problem described above, ~~and employed the same solution~~.

Notifying maintainers: @joachifm @np @prusnak @chewblacka

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

More: Tested few `electrum` functionality

- [x] `electrum` starts
- [x] `electrum` lists known transaction
- [x] `electrum` lists known addresses
- [x] "Electrum" shows up in the Xfce start menu, which starts `electrum` as expected

`nixpkgs-review` tries to rebuild 2820 packages and was therefore skipped.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

prusnak commented:
posPatches patches contrib/requirements/requirements.txt which you claim does not exist

is that true?

Yarny0 commented:
> posPatches patches contrib/requirements/requirements.txt which you claim does not exist
> 
> is that true?

Oh damn, you're absolutely right, @prusnak   I'm not sure how I grew the impression that this file does not exist in the tarball.  It does!  But the patch cannot be applied to that file -- likely due to https://github.com/spesmilo/electrum/commit/1ee6361c7e2586b74e9680835cc51cd1ad993f81 which the patch is based upon and which changes a line nearby.

I tried to apply that conflicting commit as well, but it caused other conflicts.

I guess that the `requirements.txt` file in `contrib` directory does not affect the functionality of `electrum`, so I suggest to still simply exclude that file from the patch.  Alternatively we could include our own patch.

I've nevertheless updated the branch so the comment for the excludes directive is correct.

What do you think?

prusnak commented:
> I guess that the `requirements.txt` file in `contrib` directory does not affect the functionality of `electrum`, so I suggest to still simply exclude that file from the patch. Alternatively we could include our own patch.

It is opened from setup.py to install dependencies.

I suggest you just drop the `fetchpatch` and use substituteInPlace to update `./contrib/requirements/requirements.txt` and `electrum/electrum`. While you are at it, I suggest to replace `--replace` with `--replace` in every call of `substituteInPlace`

Yarny0 committed:
Revert "electrum: fixup Exec lines in .desktop"

This reverts commit a879c725525b2321f505519ff0c71f182518f438.

It is no longer needed,
as electrum removed the PATH manipulation:
https://github.com/spesmilo/electrum/commit/f15abd7cbd55c82833ad621754d3420895414f75

However, this commit also deviates from a plain revert,
as it adds an additional replacement command to still
substitute the `--testnet` command line, as it was done before.

Note:
Those `--replace` calls are removed here in order to stricten
substitute calls with `--replace-fail` in a follow-up commit.

Yarny0 committed:
electrum: substitute with `--replace-fail`

The old `--replace` is deprecated,
we use `--replace-fail` so changes in source code that
might invalidate substitute invocations get noticed.

Cf. commit e07a2fab7f065c3fa084027f07dcf8cafbd19394 .

Yarny0 committed:
electrum: patch for aiorpcx 0.24 compatibility

electrum is quite strict about the aiorpcx version is accepts.
With commit 4bbf5300d44ee9420a59650ba668f70e3f101bde ,

Upstream already has a patch to extend
the range of aiorpcx versions it permits:
https://github.com/spesmilo/electrum/commit/171aa5ee5ad4e25b9da10f757d9d398e905b4945
As the patch does not cleanly apply to the
distribution tarball we're using, we instead apply the
required changes "by hand", with `substituteInPlace`.
On top of the changes adapted from the upstream commit,
we have to change the file `./electrum/electrum`:
It is not part of the upstream source repository,
but it gets created from `./run_electrum` when
upstream creates the distribution tarball.

Note that the commit at hand is similar to
0e513f0435b2edb2c53412df873b46c583b0f9b5 .

prusnak commented on pkgs/applications/misc/electrum/default.nix:
```diff
@@ -108,10 +119,7 @@ python3.pkgs.buildPythonApplication rec {
-                "Exec=$out/bin/electrum %u" \
-      --replace 'Exec=sh -c "PATH=\"\\$HOME/.local/bin:\\$PATH\"; electrum --testnet %u"' \
-                "Exec=$out/bin/electrum --testnet %u"
+      --replace-fail "Exec=electrum %u" "Exec=$out/bin/electrum %u"
```
let's also update testnet:

```suggestion
      --replace-fail "Exec=electrum %u" "Exec=$out/bin/electrum %u" \
      --replace-fail "Exec=electrum --testnet %u" "Exec=$out/bin/electrum --testnet %u" \
```

Yarny0 commented:
> I suggest you just drop the `fetchpatch` and use substituteInPlace to update `./contrib/requirements/requirements.txt` and `electrum/electrum`. While you are at it, I suggest to replace `--replace` with `--replace` in every call of `substituteInPlace`

I guess you ment `--replace-fail` and adapted the branch. It now also removes two useless `--replace` by reverting an old commit, and uses `substituteInPlace` instead of the patch file.  The changed `electrum.desktop` still works (shows up in Xfce start menu and starts electrum).

prusnak reviewed (changes requested):
all good except one more fix in the desktop file

prusnak commented:
@GrahamcOfBorg build electrum

prusnak commented:
thank you!
